### PR TITLE
Add monitoring pages

### LINKED ISSUE #16: Add monitoring pages
Currently, all the actions in the Application are controlled using the Django Admin Panel but we need views where we can monitor the application for new logs and check the response from the server. 

While creating templates please follow the practices in the current setup
- Layouts go in the root of the templates folder.
- Each application's templates are in their separate folder.
- Each application will have its own partials folder.

While creating Views please follow these practices
- Try using class-based views wherever possible.
- Don't use `View` if the customization can be done using mixins.
- Create a folder named `views` inside the cron application which have following folders.
    - `pages` - This will contain all the views which we will be working on in this issue
    - `api` - This will contain all the views which we will be working on in the upcoming issue. (For now just add a `.gitkeep` file in this folder
    - Inside each folder add a `views.py` and `urls.py`

- [ ] Create View's folder structure
- [ ] Implement Page Views
    - [ ] Application list page view with pagination - `/applications`
    - [ ] Application list page template
    - [ ] Application detail view which lists all the jobs paginated - `/applications/<id>`
    - [ ] Application detail view template
    - [ ] Jobs detail view which lists all the logs paginated - `/jobs/<id>`
    - [ ] Jobs detail view template
        - [ ] Should have an execute now button
        - [ ] Can apply filter on Logs ( status, created_at )
- [ ] Test the application locally and attach screenshots while making the final submission.

### PR DESCRIPTION
**Pull request by prateek2211**
fix #16 
* [X] Create View's folder structure
* [X] Implement Page Views
   * [X] Application list page view with pagination - /applications
   * [X] Application list page template
   * [X] Application detail view which lists all the jobs paginated - /applications/<id>
   * [X] Application detail view template
   * [X] Jobs detail view which lists all the logs paginated - /jobs/<id>
   * [X] Jobs detail view template
        * [X] Should have an execute now button
        * [ ] Can apply filter on Logs ( status, created_at )
* [X] Test the application locally and attach screenshots while making the final submission.

### CODE REVIEW

**Comment by prateek2211** [comment]
Screenshots:
Screenshot from 2020-05-25 10-58-05
Screenshot from 2020-05-25 10-59-03
Screenshot from 2020-05-25 10-59-16

**Comment by prateek2211** [comment]
# TODO:
* Add filter on log page
* Action on delete button
* Action on execute button

**Comment by prateek2211** [comment]
@jatinkatyal13 Please check if the UI and code architecture going the right way.

File: cron/views/pages/views.py [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+  model = Job
+  paginate_by = 10
+
+  def get_context_data(self, **kwargs):
+    log_queryset = self.get_object().log_set
+    object_list = log_queryset.order_by('pk')
+    success_count = log_queryset.filter(response__range=(200, 299)).count()
+    total_count = log_queryset.count()
+    success_rate = success_count / total_count * 100 if total_count != 0 else 100.0
+    context = super(JobDetailView, self).get_context_data(object_list=object_list,
```
Comment (jatinkatyal13) [inline cron/views/pages/views.py]: Nitpick, but in python3 you can simply write `super().get_context_data`

File: cron/views/pages/views.py [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+    return context
+
+
+class JobDetailView(generic.DetailView, MultipleObjectMixin):
+  model = Job
+  paginate_by = 10
+
+  def get_context_data(self, **kwargs):
+    log_queryset = self.get_object().log_set
+    object_list = log_queryset.order_by('pk')
```
Comment (jatinkatyal13) [inline cron/views/pages/views.py]: `order_by('-created_at')`

File: cron/views/pages/views.py [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+  model = Application
+  paginate_by = 10
+
+
+class ApplicationDetailView(generic.DetailView, MultipleObjectMixin):
+  model = Application
+  paginate_by = 10
+
+  def get_context_data(self, **kwargs):
+    object_list = self.get_object().job_set.order_by('pk')
```
Comment (jatinkatyal13) [inline cron/views/pages/views.py:17]: `order_by('-created_at')`

File: cron/views/pages/views.py [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+    context = super(ApplicationDetailView, self).get_context_data(object_list=object_list, **kwargs)
```
Comment (jatinkatyal13) [inline cron/views/pages/views.py]: use `super()`

File: cron/views/pages/views.py [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+
+from cron.models import Application, Job
+
+
+class ApplicationListView(generic.ListView):
+  model = Application
+  paginate_by = 10
+
+
+class ApplicationDetailView(generic.DetailView, MultipleObjectMixin):
```
Comment (jatinkatyal13) [inline cron/views/pages/views.py:12]: It won't make a difference here but Mixins should always be placed before the views

**Review by jatinkatyal13 — CHANGES_REQUESTED** [review]

**Comment by jatinkatyal13** [comment]
As far as UI is concerned it's good. We'll be needing more feedback to know what all information can be displayed to the user, until then this looks good for the MVP.

File: cron/views/pages/views.py (hunk 3 above)
Comment (prateek2211) [reply cron/views/pages/views.py:17]: Job doesn't contain field `created_at`

**Comment by prateek2211** [comment]
@jatinkatyal13 For delete and execute, do we need some POST pages endpoints or
some API endpoint and use AJAX

**Comment by jatinkatyal13** [comment]
So let's drop the delete button as it'll be handled in the admin panel and right now the UI is only for the monitoring

For execute now let's have POST API endpoint and we'll call the URL using the JQuery from the frontend.

File: cron/views/pages/views.py (hunk 3 above)
Comment (jatinkatyal13) [reply cron/views/pages/views.py:17]: oh my bad! Let's order job on pk only

**Comment by prateek2211** [comment]
@jatinkatyal13 Do I need to create execute jobs API in this PR, or will there be a separate issue for that?
Also, any design for filtering logs?

**Comment by jatinkatyal13** [comment]
E2E flow for executing job and monitoring is the what we aim in this PR and hence, you'll need to create the API as well.

For filter design, we can simply have different input fields on the top of the table that'll help filter different columns

**Comment by jatinkatyal13** [comment]
I've reviewed the execute API part and it's a job well done.

**Comment by prateek2211** [comment]
@jatinkatyal13 So, I guess just filter part is remaining in this PR?

**Comment by prateek2211** [comment]
@jatinkatyal13 I have added filter in logs page.

Sorry for the delay, because I was sick

**Review by jatinkatyal13 — APPROVED** [review]

**Pull request merged**